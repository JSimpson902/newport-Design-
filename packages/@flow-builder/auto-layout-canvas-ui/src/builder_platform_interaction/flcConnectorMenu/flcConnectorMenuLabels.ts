// @ts-nocheck
import menuHeader from '@salesforce/label/AlcConnectorContextualMenu.menuHeader';
import pasteSectionLabel from '@salesforce/label/AlcConnectorContextualMenu.pasteSectionLabel';
import pasteItemLabel from '@salesforce/label/AlcConnectorContextualMenu.pasteItemLabel';
import pasteItemDescription from '@salesforce/label/AlcConnectorContextualMenu.pasteItemDescription';

const mergePathSectionLabel = 'Path Actions';
const mergePathItemLabel = 'Merge with existing path';
const mergePathItemDescription = 'Delete the end element and merge this path';

export const LABELS = {
    menuHeader,
    pasteSectionLabel,
    pasteItemLabel,
    pasteItemDescription,
    mergePathSectionLabel,
    mergePathItemLabel,
    mergePathItemDescription
};
