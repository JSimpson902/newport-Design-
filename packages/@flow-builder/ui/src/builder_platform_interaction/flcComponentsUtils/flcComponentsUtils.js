import { classSet } from 'lightning/utils';
import { getStyleFromGeometry } from 'builder_platform_interaction/flowUtils';

const CLASS_MENU_OPENED = 'menu-opened';
const CLASS_IS_NEW = 'is-new';

/**
 * Utils functions providing the data needed to render flc components
 */

/**
 * Get the data needed to render a flcConnector component
 *
 * @param {ConnectorRenderInfo} connectorInfo - Info about a connector
 * @returns {Object} The flcConnector component data
 */
export function getFlcConnectorData(connectorInfo) {
    const { key, geometry } = connectorInfo;

    return {
        key,
        connectorInfo,
        style: getStyleFromGeometry(geometry),
        className: ''
    };
}

/**
 * Get the data needed to render a flcFlow component
 *
 * @param {FlowRenderInfo} flowInfo - Info about a flow
 * @returns {Object} The flcFlow component data
 */
export function getFlcFlowData(flowInfo) {
    const { key, geometry } = flowInfo;

    return {
        key,
        flowInfo,
        style: getStyleFromGeometry(geometry),
        className: ''
    };
}

/**
 * Get the data needed to render a flcCompoundNode component
 *
 * @param {NodeRenderInfo} nodeInfo - Info about a node
 * @returns {Object} The flcCompoundNode component data
 */
export function getFlcCompoundNodeData(nodeInfo) {
    const { key, geometry } = nodeInfo;

    const className = classSet({ [CLASS_IS_NEW]: nodeInfo.isNew });

    return {
        key,
        nodeInfo,
        style: getStyleFromGeometry(geometry),
        className
    };
}

/**
 * Get the data needed to render a flcNode component
 *
 * @param {NodeRenderInfo} nodeInfo - Info about a node
 * @returns {Object} The flcNode component data
 */
export function getFlcNodeData(nodeInfo) {
    const { key, menuOpened, isNew } = nodeInfo;

    const className = classSet({
        [CLASS_MENU_OPENED]: menuOpened,
        [CLASS_IS_NEW]: isNew
    }).toString();

    return {
        key,
        nodeInfo,
        className,
        style: ''
    };
}
