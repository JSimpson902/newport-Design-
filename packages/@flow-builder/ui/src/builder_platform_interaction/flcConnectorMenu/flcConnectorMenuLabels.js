import menuHeader from '@salesforce/label/AlcConnectorContextualMenu.menuHeader';
import pasteSectionLabel from '@salesforce/label/AlcConnectorContextualMenu.pasteSectionLabel';
import pasteItemLabel from '@salesforce/label/AlcConnectorContextualMenu.pasteItemLabel';
import pasteItemDescription from '@salesforce/label/AlcConnectorContextualMenu.pasteItemDescription';

export const LABELS = {
    menuHeader,
    pasteSectionLabel,
    pasteItemLabel,
    pasteItemDescription
};
