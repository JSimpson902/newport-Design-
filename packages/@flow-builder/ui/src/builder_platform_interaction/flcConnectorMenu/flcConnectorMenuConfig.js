// @ts-nocheck
import { generateGuid } from 'builder_platform_interaction/storeLib';
import { ELEMENT_TYPE } from 'builder_platform_interaction/flowMetadata';
import { LABELS } from './flcConnectorMenuLabels';

export const PASTE_ACTION = 'Paste';

export const pasteSection = {
    guid: generateGuid(),
    heading: '',
    items: [
        {
            guid: generateGuid(),
            description: LABELS.pasteItemDescription,
            icon: 'standard:record',
            iconClass: 'paste-icon',
            label: LABELS.pasteItemLabel,
            elementType: PASTE_ACTION
        }
    ],
    label: LABELS.pasteSectionLabel,
    separator: true
};

/**
 * Create FLC menu configuration from the elements metadata
 * @param {Object} elementsMetadata
 * @param {Boolean} showEndElement
 * @param {Boolean} isPasteAvailable
 */
export const configureMenu = (elementsMetadata = [], showEndElement, isPasteAvailable) => {
    const sectionDefinitionsMap = {};

    const sections = elementsMetadata.reduce(
        (acc, { section = null, description, icon, label, elementType }) => {
            if (section == null || (elementType === ELEMENT_TYPE.END_ELEMENT && !showEndElement)) {
                return acc;
            }

            let sectionDefinition = sectionDefinitionsMap[section];
            if (!sectionDefinition) {
                sectionDefinitionsMap[section] = sectionDefinition = {
                    guid: generateGuid(),
                    heading: section,
                    label: section,
                    items: []
                };

                acc.push(sectionDefinition);
            }

            sectionDefinition.items.push({
                guid: generateGuid(),
                description,
                label,
                elementType,
                icon
            });

            return acc;
        },
        isPasteAvailable ? [pasteSection] : []
    );

    return { sections };
};
