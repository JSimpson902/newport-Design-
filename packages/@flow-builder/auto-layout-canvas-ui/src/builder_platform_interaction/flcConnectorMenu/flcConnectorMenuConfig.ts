// @ts-nocheck
import { ElementType } from 'builder_platform_interaction/autoLayoutCanvas';
import { generateGuid } from 'builder_platform_interaction/storeLib';

import { ICON_SHAPE } from 'builder_platform_interaction/flcComponentsUtils';
import { LABELS } from './flcConnectorMenuLabels';

export const PASTE_ACTION = 'Paste';
export const MERGE_PATH_ACTION = 'mergePath';

export const pasteSection = {
    guid: generateGuid(),
    heading: '',
    items: [
        {
            guid: generateGuid(),
            description: LABELS.pasteItemDescription,
            icon: 'standard:record',
            iconContainerClass: 'slds-media__figure slds-listbox__option-icon',
            iconClass: 'paste-icon',
            iconSize: 'small',
            iconVariant: '',
            label: LABELS.pasteItemLabel,
            elementType: PASTE_ACTION
        }
    ],
    label: LABELS.pasteSectionLabel,
    separator: true
};

export const mergePathSection = {
    guid: generateGuid(),
    heading: LABELS.mergePathSectionLabel,
    items: [
        {
            guid: generateGuid(),
            description: LABELS.mergePathItemDescription,
            icon: 'standard:branch_merge',
            iconContainerClass: 'slds-media__figure slds-listbox__option-icon',
            iconClass: 'branch-merge',
            iconSize: 'small',
            iconVariant: '',
            label: LABELS.mergePathItemLabel,
            elementType: MERGE_PATH_ACTION
        }
    ],
    label: LABELS.mergePathSectionLabel
};

/**
 * Create FLC menu configuration from the elements metadata
 * @param {Object} elementsMetadata
 * @param {Boolean} showEndElement - Whether to show the end element item
 * @param {Boolean} canMergePath - Whether to show the merge path item
 * @param {Boolean} isPasteAvailable - If paste is available
 */
export const configureMenu = (elementsMetadata, showEndElement, isPasteAvailable, canMergePath) => {
    const sectionDefinitionsMap = {};

    const extraSections = isPasteAvailable ? [pasteSection] : [];
    if (canMergePath) {
        extraSections.push(mergePathSection);
    }

    const sections = elementsMetadata.reduce(
        (acc, { section = null, description, icon, iconShape, iconBackgroundColor, label, elementType, type }) => {
            if (section == null || (type === ElementType.END && !showEndElement)) {
                return acc;
            }

            let sectionDefinition = sectionDefinitionsMap[section];
            if (!sectionDefinition) {
                sectionDefinitionsMap[section] = sectionDefinition = {
                    guid: generateGuid(),
                    heading: section,
                    label: section,
                    items: []
                };

                acc.push(sectionDefinition);
            }

            let iconContainerClass = 'slds-media__figure slds-listbox__option-icon';
            let iconClass = '';
            let iconSize = 'small';
            let iconVariant = '';

            if (iconBackgroundColor) {
                iconClass = iconBackgroundColor;
            }

            if (iconShape === ICON_SHAPE.CIRCLE && type === ElementType.END) {
                iconClass = `${iconClass} end-element-svg`;
                iconSize = 'xx-small';
                iconVariant = 'inverse';
            } else if (iconShape === ICON_SHAPE.DIAMOND) {
                iconContainerClass = `${iconContainerClass} rotate-icon-container slds-icon-standard-decision`;
                iconClass = `${iconClass} rotate-icon-svg`;
            }

            sectionDefinition.items.push({
                guid: generateGuid(),
                description,
                label,
                elementType,
                icon,
                iconContainerClass,
                iconClass,
                iconSize,
                iconVariant
            });

            return acc;
        },
        extraSections
    );

    return { sections };
};
