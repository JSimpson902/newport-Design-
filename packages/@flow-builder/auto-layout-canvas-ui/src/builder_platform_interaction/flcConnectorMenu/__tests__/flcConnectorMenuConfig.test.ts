// @ts-nocheck
import { configureMenu } from '../flcConnectorMenuConfig';
import { ICON_SHAPE } from 'builder_platform_interaction/flcComponentsUtils';
import { LABELS } from '../flcConnectorMenuLabels';

jest.mock('builder_platform_interaction/sharedUtils', () => {
    return {
        storeUtils: {
            generateGuid: jest.fn(() => 1)
        }
    };
});

const metaData = [
    {
        section: 'Interaction',
        type: 'default',
        icon: 'standard:screen',
        label: 'Screen',
        value: 'Screen',
        elementType: 'Screen',
        description: 'Collect information from',
        isSupported: true
    },
    {
        section: 'Logic',
        type: 'branch',
        icon: 'standard:decision',
        iconShape: ICON_SHAPE.DIAMOND,
        label: 'Decision',
        value: 'Decision',
        elementType: 'Decision',
        description: 'Create Decision',
        isSupported: true
    },
    {
        section: 'Logic',
        type: 'end',
        icon: 'utility:stop',
        iconBackgroundColor: 'background-red',
        iconShape: ICON_SHAPE.CIRCLE,
        label: 'End',
        value: 'End',
        elementType: 'End',
        description: 'Create End',
        isSupported: true
    }
];

describe('connector menu config', () => {
    it('Match Snapshot with both paste and merge option', () => {
        expect(configureMenu(metaData, false, true, true)).toMatchSnapshot();
    });

    it('actionSection should have the right label', () => {
        expect(configureMenu(metaData, false, true, false).sections[0].label).toBe(LABELS.actionSectionLabel);
    });

    it('pasteItem should have the right label', () => {
        expect(configureMenu(metaData, false, true, false).sections[0].items[0].label).toBe(LABELS.pasteItemLabel);
    });

    it('pasteSection item should have the right icon', () => {
        expect(configureMenu(metaData, false, true, false).sections[0].items[0].icon).toBe('utility:paste');
    });

    it('pasteSection item should have the right rowClass', () => {
        expect(configureMenu(metaData, false, true, false).sections[0].items[0].rowClass).toBe(
            'slds-listbox__item action-row-line-height'
        );
    });

    it('mergePath should have the right label', () => {
        expect(configureMenu(metaData, false, true, true).sections[0].items[1].label).toBe(LABELS.mergePathItemLabel);
    });

    it('mergePath item should have the right icon', () => {
        expect(configureMenu(metaData, false, true, true).sections[0].items[1].icon).toBe('utility:merge');
    });

    it('mergePath item should have the right icon class', () => {
        expect(configureMenu(metaData, false, true, true).sections[0].items[1].iconClass).toBe('branch-merge');
    });

    it('mergePath item should have the right rowClass', () => {
        expect(configureMenu(metaData, false, true, true).sections[0].items[1].rowClass).toBe(
            'slds-listbox__item action-row-line-height'
        );
    });

    it('gergePath should have the right label', () => {
        expect(configureMenu(metaData, false, true, true, true, false).sections[0].items[0].label).toBe(
            LABELS.gergePathItemLabel
        );
    });

    it('gergePath item should have the right icon', () => {
        expect(configureMenu(metaData, false, true, true, true, false).sections[0].items[0].icon).toBe(
            'utility:level_down'
        );
    });

    it('gergePath item should have the right rowClass', () => {
        expect(configureMenu(metaData, false, true, true, true, false).sections[0].items[0].rowClass).toBe(
            'slds-listbox__item action-row-line-height'
        );
    });

    it('reRoute/delete should be displayed and Gerge should not', () => {
        const menu = configureMenu(metaData, false, true, true, true, true);
        expect(menu.sections[0].items[0].label).toBe(LABELS.reRouteGergePathItemLabel);
        expect(menu.sections[0].items[1].label).toBe(LABELS.deleteGergePathItemLabel);
        expect(menu.sections[0].items[2].label).toBe(LABELS.pasteItemLabel);
    });

    it('reRoutePath item should have the right icon', () => {
        expect(configureMenu(metaData, false, true, true, true, true).sections[0].items[0].icon).toBe(
            'utility:level_down'
        );
    });

    it('reRoutePath item should have the right rowClass', () => {
        expect(configureMenu(metaData, false, true, true, true, true).sections[0].items[0].rowClass).toBe(
            'slds-listbox__item action-row-line-height'
        );
    });

    it('deletePath item should have the right icon', () => {
        expect(configureMenu(metaData, false, true, true, true, true).sections[0].items[1].icon).toBe('utility:delete');
    });

    it('deletePath item should have the right rowClass', () => {
        expect(configureMenu(metaData, false, true, true, true, true).sections[0].items[1].rowClass).toBe(
            'slds-listbox__item action-row-line-height'
        );
    });
});
