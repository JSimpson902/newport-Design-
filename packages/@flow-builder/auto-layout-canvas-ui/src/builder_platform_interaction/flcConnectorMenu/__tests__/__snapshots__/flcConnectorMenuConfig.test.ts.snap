// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connector menu config should have the reconnect branch menu item 1`] = `
Object {
  "sections": Array [
    Object {
      "guid": 1,
      "heading": "Path Actions",
      "items": Array [
        Object {
          "description": "Delete the end element and merge this path",
          "elementType": "mergePath",
          "guid": 1,
          "icon": "standard:branch_merge",
          "iconClass": "branch-merge",
          "iconContainerClass": "slds-media__figure slds-listbox__option-icon",
          "iconSize": "small",
          "iconVariant": "",
          "label": "Merge with existing path",
        },
      ],
      "label": "Path Actions",
    },
    Object {
      "guid": 1,
      "heading": "Interaction",
      "items": Array [
        Object {
          "description": "Collect information from",
          "elementType": "Screen",
          "guid": 1,
          "icon": "standard:screen",
          "iconClass": "",
          "iconContainerClass": "slds-media__figure slds-listbox__option-icon",
          "iconSize": "small",
          "iconVariant": "",
          "label": "Screen",
        },
      ],
      "label": "Interaction",
    },
    Object {
      "guid": 1,
      "heading": "Logic",
      "items": Array [
        Object {
          "description": "Create Decision",
          "elementType": "Decision",
          "guid": 1,
          "icon": "standard:decision",
          "iconClass": " rotate-icon-svg",
          "iconContainerClass": "slds-media__figure slds-listbox__option-icon rotate-icon-container slds-icon-standard-decision",
          "iconSize": "small",
          "iconVariant": "",
          "label": "Decision",
        },
      ],
      "label": "Logic",
    },
  ],
}
`;
